### ISSUE TITLE
Proposal: Allow the App to Print without showing the Confirmation Dialog to user

### ISSUE THREAD

**Issue by TonyHenrique** [microsoft/ProjectReunion#124]
# Proposal: Allow the App to Print without showing the Confirmation Dialog to user

## Summary

We should be able to programatically list printers, select the printer, configure paper size, orientation, settings... and PRINT without user interaction.

## Rationale
I was trying to make a Touch Screen Kiosk Receipt Print application on UWP, but the samples show

`await Windows.Graphics.Printing.PrintManager.ShowPrintUIAsync();`
This brings the UWP Print Dialog.

I need to print directly without this dialog. How to do this?

I need something like

`await Windows.Graphics.Printing.PrintManager.PrintAsync();`
We should be able to programatically list printers, select the printer, configure paper size, orientation, settings...

This will enable:
* Enable easy printing from Voice / Touch applications
* Make the printing API more complete, like the WPF Printing API was.
* This Permission will be very useful for **ATM Applications**, **Touch Screen Kiosk Apps**, **Barcode Printing Software**, **Point of Sale applications**, **Line-of-Business Applications**, etc;

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers Print without user interaction. | Must |
| This proposal will allow the App to Configure Paper Size  | Must |
| This proposal will allow the App to Set Page Orientation  | Must |
| This proposal will allow the App to List printers | Must |
| This proposal will allow end users to give the Printing without interaction permission on app install  (**AllowPrintingWithoutUserIteraction**)| Should |
| This proposal will allow the App to Change the Default Printer | Could |
| This proposal will allow the App to set Printing Settings like Color, Grayscale, Draft | Could |

## Important Notes
This may require the creation of a new permission/capability for **UWP AppModel**:
**AllowPrintingWithoutUserIteraction**

**Comment by jonwis** [comment]
This is very similar to existing issue #52, but has more detail.  @awsomecode would you mind if we made this the primary since it has the "proposal" format with details?  What else would you want to add to the proposal?

**Comment by AwsomeCode** [comment]
@jonwis Ya, this one will cover my requirement as well. You can make this primary.

**Comment by ptorr-msft** [comment]
@TonyHenrique FYI we do not show consent prompts on install[1]; we show them on first-use (more correctly: the first time the app requests the permission, which might be before it actually "uses" it, depending on how the app is written and the specific API in question). Otherwise, agreed that this feature would need a permission that the user could consents to and can later change their mind in Settings.

[1] We do perform a UAC prompt for anything that impacts the machine / other users during install.

**Comment by stevewri** [comment]
Moving to backlog for future potential feature work.

**Comment by AwsomeCode** [comment]
@stevewri This wont be available on WinUI3 ?

**Comment by ptorr-msft** [comment]
> This wont be available on WinUI3 ?

Printing without confirmation requires work at the security / isolation level, not just the UI level.

**Comment by AwsomeCode** [comment]
This is so much import for retail softwares. Without this we can't convert WPF apps to UWP/WinUI3

**Comment by Felix-Dev** [comment]
@AwsomeCode You should be able to print without an UI confirmation dialog if you will port your WPF apps to WinUI 3 *desktop* apps. @ptorr-msft's comment is about WinUI 3 UWP apps (since the requested change will require changes in the UWP app model implementation and that's not something WinUI is responsible for).

**Comment by Rothanan** [comment]
I'm trying to build a POS application with UWP, can't get the POS to recognize the cheaper thermal printers out there, I'm waiting on this to be realeased so I can properly deal with the printers that are not supported by the POS namespace.

**Comment by ptorr-msft** [comment]
@Rothanan can you share the details of the printers that don't work? Are they specifically sold as POS devices that adhere to the POS standards?

**Comment by AwsomeCode** [comment]
@ptorr-msft this are the few I have tried.
https://www.atpos.in/product/atpos-usb-bluetooth-e-200-portable-thermal-receipt-printer/
https://www.atpos.in/product/atpos-hop-h58-58mm-thermal-receipt-printer/

Surprisingly they work with Android.

**Comment by ysc3839** [comment]
I wonder how macOS deal with such issue. Is there a "print permission", or apps are allowed to print without confirm?

**Comment by Rothanan** [comment]
@ptorr-msft they are very basic printers, they only work through ESC/POS. I can get them working on android, but not in UWP:

Here's a model:
https://www.amazon.com/Wireless-Bluetooth-MUNBYN-Portable-Compatible/dp/B07N86DR4J

They are sold mainly as mobile bluetooth thermal printers(they also come with an usb conection) and because they are cheaper most customers wanna buy and use them.

**Comment by mdtauk** [comment]
POS printers, sounds like an IOT device situation, and that supports device drivers right?  Which would allow direct talking to the printer device right?

**Comment by Rothanan** [comment]
Yeah, you're right, in android devices, all I do is send ESC/POS commands directly to the printer.

 I'm looking for a way to do the same in UWP.

**Comment by ptorr-msft** [comment]
@Rothanan thanks for the info; I will send to the POS team internally to see if they know of any issues with these devices.

**Comment by Rothanan** [comment]
@ptorr-msft, np.

I hope they can do something about it, looking forward for some good news.

**Comment by AwsomeCode** [comment]
@ptorr-msft any update from pos team?

**Comment by ptorr-msft** [comment]
Sorry @AwsomeCode , some rather urgent personal stuff came up and I've been distracted. Tracking them down now and will pass along. Thanks for the reminder.

**Comment by Rothanan** [comment]
Sorry to bother, but, any news on this?

**Comment by ptorr-msft** [comment]
Can you confirm if you are using the POS APIs or the regular printing APIs?

**Comment by AwsomeCode** [comment]
**In UWP** 
Its not POS APIs or Regular Printing API.
I have tried [RawPrint](https://github.com/frogmorecs/RawPrint)
Which sends RAW ESC/POS data commands to printer and it print silently on printer. It works in debug mode. But on published app its not allowed.
For published app we have to use [Desktop Extension for UWP](https://stefanwick.com/2018/04/06/uwp-with-desktop-extension-part-1/). Which is not best experience in development. (Our UWP app never get approved in store. They say its crashing and In our machines its working fine.)

**In Android** 
Sending RAW ESC/POS data commands via Bluetooth and its gets print silently.

**In iOS** 
Printer is not getting connected via Bluetooth.

Below are the printers I am referring to.
https://www.atpos.in/product/atpos-usb-bluetooth-e-200-portable-thermal-receipt-printer/
https://www.atpos.in/product/atpos-hop-h58-58mm-thermal-receipt-printer/

**Comment by ptorr-msft** [comment]
> I have tried RawPrint
Which sends RAW ESC/POS data commands to printer and it print silently on printer. It works in debug mode. But on published app its not allowed.

That project says it relies on .NET 4, which isn't supported for UWP apps (only .NET Native is supported). Can you pair the printers first and then use the POS API?

**Comment by AwsomeCode** [comment]
I have tried this sample project it did not worked.
https://docs.microsoft.com/en-us/samples/microsoft/windows-universal-samples/posprinter/

**Comment by Rothanan** [comment]
> Can you confirm if you are using the POS APIs or the regular printing APIs?

I tried using the POS API, it doesn't recognize the ESC/POS printer, the POS API support very few printers which renders it unusable.

We are just looking for a away to send ESC/POS commands to ANY printer that understands it.

**Comment by Rothanan** [comment]
I did get it working in UWP using the code bellow, but it doesn't go trhough to the app store, it says I can't use these APIs.

  ```
public class RawPrintingService
    {
        // Structure and API declarions:
        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
        public class DOCINFOA
        {
            [MarshalAs(UnmanagedType.LPStr)]
            public string pDocName;
            [MarshalAs(UnmanagedType.LPStr)]
            public string pOutputFile;
            [MarshalAs(UnmanagedType.LPStr)]
            public string pDataType;
        }

        [DllImport("winspool.Drv", EntryPoint = "OpenPrinterA", SetLastError = true, CharSet = CharSet.Unicode, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool OpenPrinter([MarshalAs(UnmanagedType.LPStr)] string szPrinter, out IntPtr hPrinter, IntPtr pd);

        [DllImport("winspool.Drv", EntryPoint = "ClosePrinter", SetLastError = true, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool ClosePrinter(IntPtr hPrinter);

        [DllImport("winspool.Drv", EntryPoint = "StartDocPrinterA", SetLastError = true, CharSet = CharSet.Unicode, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool StartDocPrinter(IntPtr hPrinter, Int32 level, [In, MarshalAs(UnmanagedType.LPStruct)] DOCINFOA di);

        [DllImport("winspool.Drv", EntryPoint = "EndDocPrinter", SetLastError = true, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool EndDocPrinter(IntPtr hPrinter);

        [DllImport("winspool.Drv", EntryPoint = "StartPagePrinter", SetLastError = true, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool StartPagePrinter(IntPtr hPrinter);

        [DllImport("winspool.Drv", EntryPoint = "EndPagePrinter", SetLastError = true, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool EndPagePrinter(IntPtr hPrinter);

        [DllImport("winspool.Drv", EntryPoint = "WritePrinter", SetLastError = true, ExactSpelling = true, CallingConvention = CallingConvention.StdCall)]
        public static extern bool WritePrinter(IntPtr hPrinter, IntPtr pBytes, Int32 dwCount, out Int32 dwWritten);

        // SendBytesToPrinter()
        // When the function is given a printer name and an unmanaged array
        // of bytes, the function sends those bytes to the print queue.
        // Returns true on success, false on failure.
        public static bool SendBytesToPrinter(string szPrinterName, IntPtr pBytes, Int32 dwCount)
        {
            Int32 dwError = 0, dwWritten = 0;
            IntPtr hPrinter = new IntPtr(0);
            DOCINFOA di = new DOCINFOA();
            bool bSuccess = false; // Assume failure unless you specifically succeed.

            di.pDocName = "RAW Document";
            // Win7
            di.pDataType = "RAW";

            // Win8+
            // di.pDataType = "XPS_PASS";

            // Open the printer.
            if (OpenPrinter(szPrinterName.Normalize(), out hPrinter, IntPtr.Zero))
            {
                // Start a document.
                if (StartDocPrinter(hPrinter, 1, di))
                {
                    // Start a page.
                    if (StartPagePrinter(hPrinter))
                    {
                        // Write your bytes.
                        bSuccess = WritePrinter(hPrinter, pBytes, dwCount, out dwWritten);
                        EndPagePrinter(hPrinter);
                    }
                    EndDocPrinter(hPrinter);
                }
                ClosePrinter(hPrinter);
            }
            // If you did not succeed, GetLastError may give more information
            // about why not.
            if (bSuccess == false)
            {
                dwError = Marshal.GetLastWin32Error();
            }
            return bSuccess;
        }

        public static bool SendFileToPrinter(string szPrinterName, string szFileName)
        {
            // Open the file.
            FileStream fs = new FileStream(szFileName, FileMode.Open);
            // Create a BinaryReader on the file.
            BinaryReader br = new BinaryReader(fs);
            // Dim an array of bytes big enough to hold the file's contents.
            Byte[] bytes = new Byte[fs.Length];
            bool bSuccess = false;
            // Your unmanaged pointer.
            IntPtr pUnmanagedBytes = new IntPtr(0);
            int nLength;

            nLength = Convert.ToInt32(fs.Length);
            // Read the contents of the file into the array.
            bytes = br.ReadBytes(nLength);
            // Allocate some unmanaged memory for those bytes.
            pUnmanagedBytes = Marshal.AllocCoTaskMem(nLength);
            // Copy the managed byte array into the unmanaged array.
            Marshal.Copy(bytes, 0, pUnmanagedBytes, nLength);
            // Send the unmanaged bytes to the printer.
            bSuccess = SendBytesToPrinter(szPrinterName, pUnmanagedBytes, nLength);
            // Free the unmanaged memory that you allocated earlier.
            Marshal.FreeCoTaskMem(pUnmanagedBytes);
            fs.Close();
            fs.Dispose();
            fs = null;
            return bSuccess;
        }
        public static bool SendStringToPrinter(string szPrinterName, string szString)
        {
            IntPtr pBytes;
            Int32 dwCount;
            // How many characters are in the string?
            dwCount = szString.Length;
            // Assume that the printer is expecting ANSI text, and then convert
            // the string to ANSI text.
            pBytes = Marshal.StringToCoTaskMemAnsi(szString);
            // Send the converted ANSI string to the printer.
            SendBytesToPrinter(szPrinterName, pBytes, dwCount);
            Marshal.FreeCoTaskMem(pBytes);
            return true;
        }
    }
```

**Comment by Rothanan** [comment]
I had to create a WPF app just to receive a URI with a path to a text file and them print it.

**Comment by AwsomeCode** [comment]
@ptorr-msft any update from the pos team?

**Comment by ptorr-msft** [comment]
Sorry, I've been out of the office. I got this link: 

https://docs.microsoft.com/en-us/windows/uwp/devices-sensors/pos-device-support#receipt-printer-posprinter

**Comment by AwsomeCode** [comment]
Can pos team provided any sample for 
MUNBYN Android Bluetooth Mobile Thermal Receipt Printer, MUNBYN 2 Inches 58MM Impresora térmica Printer with Leather Belt for Business ESC/POS, Does NOT Support iOS Devices,Square https://www.amazon.com/dp/B0753B5KYG/ref=cm_sw_r_cp_api_fabt1_IN-TFbH0SH19A

**Comment by Rothanan** [comment]
> Sorry, I've been out of the office. I got this link: 
> 
> https://docs.microsoft.com/en-us/windows/uwp/devices-sensors/pos-device-support#receipt-printer-posprinter

Thanks for the link, but, we've all stumbled upon this link and it does nothing for what we are trying to achieve.

We just wanna be able to silently print stuff, if for that we need to request permission for the user, so be it. 

It would be awesome to also be able to send esc/pos commands directly to the printer, just like in the code I posted above. It works on debug mode but it doesn't pass to the Microsoft Store because it says I'm using APIs that are not always available.

**Comment by AwsomeCode** [comment]
@ptorr-msft I also have same situation as Rothanan

**Comment by AwsomeCode** [comment]
@ptorr-msft can someone from the POS team can post on this issue?

**Comment by ptorr-msft** [comment]
Since this is now about POS printing, and not general consumer printing without prompts, can you start a new issue?

**Comment by TonyHenrique** [comment]
In my case, I was planning to use general consumer printers from UWP without the user prompts for each printing job, using something like a **PrintAsync** instead of **ShowPrintUIAsync** . It would be OK if this may require the creation of a new permission/capability for UWP AppModel like: **AllowPrintingWithoutUserIteraction**

**Comment by Rothanan** [comment]
I'm with @TonyHenrique, I need that too.

**Comment by DoubleDBE** [comment]
Hi everyone,

We develop POS software and we heavily rely on the printing options in WPF.

For example we create our layouts in XAML for receipt tickets / invoices / labels etc, even almost everything if I think about it.

We primarily use the PrintVisual method that we can use in WPF to print our XAML layouts to all kinds of printers.

Due to the nature of our clients (restaurants etc) they need to be able to quickly print a receipt without being bothered with dialogs before they can actually print something.

Is this something that wil be implemented in / ported to WinUI 3?

**Comment by mirzodalerm** [comment]
We need this too.

**Comment by wolfgang-ironsoftware** [comment]
> I wonder how macOS deal with such issue. Is there a "print permission", or apps are allowed to print without confirm?

I don't believe macOS APIs for printing require permission. Actually Apple's printing API has a boolean that indicates whether to show or hide the dialog. So just setting the boolean to false will print without showing the dialog.

**Comment by MoStAfAmAhMoUdAbDo** [comment]
i am new in flutter develop and i want to make direct print for pdf file without showing the preview of printing for the user and make this for printer device that in the same network and i have its ip address .

**Comment by castorix** [comment]
I uploaded a test  ([WinUI3_Direct2D_Print](https://github.com/castorix/WinUI3_Direct2D_Print)) with [ID2D1PrintControl](https://learn.microsoft.com/en-us/windows/win32/api/d2d1_1/nn-d2d1_1-id2d1printcontrol) which can print silently (can be done with GDI too)
(by printing a few controls using [RenderTargetBitmap](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.rendertargetbitmap?view=windows-app-sdk-1.6), which works fine when the control is not virtualized)

**Comment by RDMacLachlan** [comment]
The Win32 APIs are available to handling App to Printing. Please comment if these APIs are not sufficient, and we can re-open this issue.
